File: src/cybercity_engine/state.py
"""Runtime state manager for the CyberCity engine.

The state manager owns the in-memory ``WorldState`` and persists periodic
snapshots to PostgreSQL. It is the only component allowed to mutate runtime
state directly.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from .models import (
    EventNode,
    EventType,
    PlayerState,
    ScenarioState,
    ServiceState,
    ServiceStatus,
    TopologyGraph,
    WorldState,
)

__all__ = ["StateManager"]


class StateManager:
    """Owns the mutable runtime world state."""

    def __init__(self, topology: TopologyGraph) -> None:
        self.topology = topology
        self.world = WorldState()
        self._init_services()

    def _init_services(self) -> None:
        for svc_id in self.topology.services:
            self.world.services[svc_id] = ServiceState(service_id=svc_id)

    # ──────────────────────────────────────────────────────────────────
    # Service state mutations
    # ──────────────────────────────────────────────────────────────────
    def set_service_status(
        self, service_id: str, status: ServiceStatus, reason_event: EventNode | None = None
    ) -> EventNode | None:
        svc = self._require_service(service_id)
        if svc.status == status:
            return None

        old_status = svc.status
        svc.status = status
        svc.last_heartbeat = datetime.now(UTC)

        return EventNode(
            tick=self.world.tick,
            source_type="engine",
            source_id="state-manager",
            event_type=EventType.STATE_CHANGE,
            target_id=service_id,
            payload={
                "entity": "service",
                "field": "status",
                "old_value": old_status,
                "new_value": status,
                "parent_event_id": reason_event.event_id if reason_event else None,
            },
        )

    def record_heartbeat(self, service_id: str, timestamp: datetime | None = None) -> None:
        svc = self._require_service(service_id)
        svc.last_heartbeat = timestamp or datetime.now(UTC)

    def mark_seen_by(self, service_id: str, observer_id: str) -> None:
        svc = self._require_service(service_id)
        if observer_id not in svc.seen_by:
            svc.seen_by.append(observer_id)

    def set_health(self, service_id: str, health: float) -> None:
        svc = self._require_service(service_id)
        svc.health = max(0.0, min(1.0, health))

    def set_flag(self, service_id: str, key: str, value: Any) -> None:
        svc = self._require_service(service_id)
        svc.flags[key] = value

    # ──────────────────────────────────────────────────────────────────
    # Player / scenario state
    # ──────────────────────────────────────────────────────────────────
    def add_player(self, player: PlayerState) -> None:
        self.world.players[player.player_id] = player

    def set_scenario(self, scenario: ScenarioState | None) -> None:
        self.world.active_scenario = scenario

    # ──────────────────────────────────────────────────────────────────
    # Tick / snapshot helpers
    # ──────────────────────────────────────────────────────────────────
    def increment_tick(self) -> None:
        self.world.tick += 1

    # ──────────────────────────────────────────────────────────────────
    # Helpers
    # ──────────────────────────────────────────────────────────────────
    def _require_service(self, service_id: str) -> ServiceState:
        if service_id not in self.world.services:
            raise KeyError(f"unknown service {service_id!r}")
        return self.world.services[service_id]

File: internal/domain/state/state.go
// Package state владеет изменяемым runtime-состоянием CyberCity Engine.
package state

import (
	"time"

	"github.com/TheCipherKeeper/cybercity-engine/internal/domain/models"
	"github.com/google/uuid"
)

// StateManager — единственный владелец WorldState.
type StateManager struct {
	Topology *models.TopologyGraph
	World    *models.WorldState
}

// NewStateManager создаёт StateManager и инициализирует состояние сервисов
// по топологическому графу.
func NewStateManager(topology *models.TopologyGraph) *StateManager {
	world := &models.WorldState{
		Tick:      0,
		StartedAt: time.Now().UTC(),
		Services:  make(map[string]*models.ServiceState),
		Players:   make(map[string]*models.PlayerState),
		Variables: make(map[string]any),
	}
	sm := &StateManager{
		Topology: topology,
		World:    world,
	}
	sm.initServices()
	return sm
}

func (sm *StateManager) initServices() {
	for id := range sm.Topology.Services {
		sm.World.Services[id] = &models.ServiceState{
			ServiceID: id,
			Status:    models.ServiceStatusUp,
			Health:    1.0,
			SeenBy:    []string{},
			Flags:     make(map[string]any),
			Variables: make(map[string]any),
		}
	}
}

// SetServiceStatus меняет статус сервиса и возвращает STATE_CHANGE событие.
// Если статус не изменился, возвращает nil.
func (sm *StateManager) SetServiceStatus(serviceID string, status models.ServiceStatus, reasonEvent *models.EventNode) *models.EventNode {
	svc := sm.requireService(serviceID)
	if svc.Status == status {
		return nil
	}

	oldStatus := svc.Status
	svc.Status = status
	now := time.Now().UTC()
	svc.LastHeartbeat = &now

	payload := map[string]any{
		"entity":    "service",
		"field":     "status",
		"old_value": string(oldStatus),
		"new_value": string(status),
	}
	if reasonEvent != nil {
		payload["parent_event_id"] = reasonEvent.EventID
	}

	return &models.EventNode{
		EventID:    uuid.NewString(),
		Tick:       sm.World.Tick,
		Timestamp:  time.Now().UTC(),
		SourceType: models.SourceTypeEngine,
		SourceID:   "state-manager",
		EventType:  models.EventTypeStateChange,
		TargetID:   serviceID,
		Payload:    payload,
		Status:     models.EventStatusPending,
	}
}

// RecordHeartbeat обновляет время последнего heartbeat сервиса.
func (sm *StateManager) RecordHeartbeat(serviceID string, timestamp *time.Time) {
	svc := sm.requireService(serviceID)
	if timestamp != nil {
		svc.LastHeartbeat = timestamp
		return
	}
	now := time.Now().UTC()
	svc.LastHeartbeat = &now
}

// MarkSeenBy добавляет observer в список тех, кто видел сервис.
func (sm *StateManager) MarkSeenBy(serviceID, observerID string) {
	svc := sm.requireService(serviceID)
	for _, existing := range svc.SeenBy {
		if existing == observerID {
			return
		}
	}
	svc.SeenBy = append(svc.SeenBy, observerID)
}

// SetHealth устанавливает здоровье сервиса в диапазоне [0, 1].
func (sm *StateManager) SetHealth(serviceID string, health float64) {
	svc := sm.requireService(serviceID)
	if health < 0 {
		health = 0
	}
	if health > 1 {
		health = 1
	}
	svc.Health = health
}

// SetFlag устанавливает runtime-флаг сервиса.
func (sm *StateManager) SetFlag(serviceID, key string, value any) {
	svc := sm.requireService(serviceID)
	if svc.Flags == nil {
		svc.Flags = make(map[string]any)
	}
	svc.Flags[key] = value
}

// AddPlayer добавляет игрока в мир.
func (sm *StateManager) AddPlayer(player *models.PlayerState) {
	sm.World.Players[player.PlayerID] = player
}

// SetScenario устанавливает или сбрасывает активный сценарий.
func (sm *StateManager) SetScenario(scenario *models.ScenarioState) {
	sm.World.ActiveScenario = scenario
}

// IncrementTick увеличивает номер текущего tick.
func (sm *StateManager) IncrementTick() {
	sm.World.Tick++
}

func (sm *StateManager) requireService(serviceID string) *models.ServiceState {
	svc, ok := sm.World.Services[serviceID]
	if !ok {
		panic("unknown service " + serviceID)
	}
	return svc
}
